packages/build/src/plugins_core/deploy/buildbot_client.ts

```diff
@@ -0,0 +1,167 @@
+  return { host, port: Number(port) }
+}
+
+/**
+ * Emits the connect event
+ */
+export const connectBuildbotClient = addAsyncErrorMessage(async (client: net.Socket) => {
+  await pEvent(client, 'connect')
+}, 'Could not connect to buildbot')
+
+/**
+ * Closes the buildbot client and its connection
+ */
+export const closeBuildbotClient = async function (client: net.Socket) {
+  if (client.destroyed) {
+    return
+  }
+
+  await promisify(client.end.bind(client))()
+}
+
+const writePayload = addAsyncErrorMessage(async (buildbotClient: net.Socket, payload: BuildRequest) => {
+  await promisify(buildbotClient.write.bind(buildbotClient))(JSON.stringify(payload))
+}, 'Could not send payload to buildbot')
+
+const getNextParsedResponsePromise = addAsyncErrorMessage<BuildbotResponse>(async (buildbotClient: net.Socket) => {
+  const data = await pEvent<'data', string>(buildbotClient, 'data')
+  return JSON.parse(data) as BuildbotResponse
+}, 'Invalid response from buildbot')
+
+/**
+ * Initates the deploy with the given buildbot client
+ */
+export const deploySiteWithBuildbotClient = async function ({ client, events, buildDir, repositoryRoot, constants }) {
+  const action = shouldWaitForPostProcessing(events) ? Action.DeploySiteAndWait : Action.DeploySite
+  const deployDir = getDeployDir({ buildDir, repositoryRoot, constants })
+  const payload = { action, deployDir }
+
+  const [response] = await Promise.all([getNextParsedResponsePromise(client), writePayload(client, payload)])
+
+  if (!response.succeeded) {
+    const { error, code, error_type } = response?.values || {}
+    return handleDeployError(error, code, error_type)
+  }
+}
+
+/**
+ * The file paths in the buildbot are relative to the repository root.
+ * However, the file paths in Build plugins, including `constants.PUBLISH_DIR`
+ * are relative to the build directory, which is different when there is a
+ * base directory. This converts it.
+ * We need to call `normalize()` in case the publish directory is the
+ * repository root, so `deployDir` is "." not ""
+ */
+const getDeployDir = function ({ buildDir, repositoryRoot, constants: { PUBLISH_DIR } }) {
+  const absolutePublishDir = resolve(buildDir, PUBLISH_DIR)
+  const relativePublishDir = relative(repositoryRoot, absolutePublishDir)
+  const deployDir = normalize(relativePublishDir)
+  return deployDir
+}
+
+/**
+ * We distinguish between user errors and system errors during deploys
+ */
+const handleDeployError = function (error?: string, errorCode?: string, errorType?: ErrorType) {
+  if (errorCode !== undefined) {
+    const err = new Error(`Deploy did not succeed with HTTP Error ${errorCode}: ${error}`)
+    if (errorCode.startsWith('5')) {
+      const errorInfo = { type: 'deployInternal', location: { statusCode: errorCode } }
+      addErrorInfo(err, errorInfo)
+      throw err
+    }
+    const errorInfo = { type: 'deploy', location: { statusCode: errorCode } }
+    addErrorInfo(err, errorInfo)
+    throw err
+  }
```

Comment:
Other non 4xx errors will get reported as user errors.

Before we had custom logic to cater for 422's (which was never executed because the `code` was never populated):
- https://github.com/netlify/build/blob/5eaf9de39853307607b8882ba7a47aef9aa083bf/packages/build/src/plugins_core/deploy/buildbot_client.js#L89-L95
- https://cloud.us.humio.com/netlify-us-production/search?live=false&query=%24bbot()%20%7C%20%22File%20upload%20failed%20because%20of%20mismatched%20SHAs.%22&start=30d&tz=Europe%2FLisbon

We can't do this any longer because this is no longer accurate. Our API returns 422's for multiple scenarios (invalid function names, functions outside of the customer plan, builds that have already finished, etc.), see some examples:
- https://ui.honeycomb.io/gatsbyinc/environments/buildbot-production/datasets/-netlify-build/result/9YM7TQGXKTf

If this ends up being an issue in the future we can look into putting in place better messaging around these errors. For now I think this is already a nice improvement.